internal/app/machined/pkg/controllers/runtime/booted_entry.go

```diff
@@ -0,0 +1,147 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+
+	machineruntime "github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
+	"github.com/siderolabs/talos/internal/app/machined/pkg/runtime/v1alpha1/bootloader/sdboot"
+	runtimeres "github.com/siderolabs/talos/pkg/machinery/resources/runtime"
+	"github.com/siderolabs/talos/pkg/machinery/resources/v1alpha1"
+)
+
+// BootedEntryController is a controller that updates the booted entry resource.
+type BootedEntryController struct {
+	V1Alpha1Mode machineruntime.Mode
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *BootedEntryController) Name() string {
+	return "runtime.BootedEntryController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *BootedEntryController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: v1alpha1.NamespaceName,
+			Type:      runtimeres.SecurityStateType,
+			ID:        optional.Some(runtimeres.SecurityStateID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *BootedEntryController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: runtimeres.BootedEntryType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *BootedEntryController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// If we're booted in Container mode, short-circuit the controller.
+	if ctrl.V1Alpha1Mode == machineruntime.ModeContainer {
+		return nil
+	}
+
+	for {
```

Comment:
Maybe this could just run one-shot somehow? Just wait for the first SecurityState appearing, and then create the resource and exit?